### PR TITLE
Add Expr::column_refs to find column references without copying

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of  https://github.com/apache/datafusion/issues/10505

## Rationale for this change

Code that uses `Expr::to_columns` or `expr_to_columns` copies the column name strings, often to simply check if a column is present. A non copying API would be more efficient, and now thanks to the improvements in the TreeNode API (from https://github.com/apache/datafusion/pull/10543, thanks @peter-toth!) it is straightforward to implement

## What changes are included in this PR?

1. Add `Expr::column_refs` and `Expr::add_column_refs`
2. Updated some places in the code to use this new API. I will make a follow on ticket to deprecate the old apis and update code to use it
3. Deprecate `exprlist_to_columns` which is not used anywhere in the datafusion codebase

I plan to deprecate `Expr::to_columns` and `expr_to_columns` as follow on PRs

## Are these changes tested?

Yes, by existing tests and new doc examples

Performance results show basically no improvement (maybe it gets faster for some of the queries with larger numbers of columns) but it may also be noise

<details><summary>Details</summary>
<p>

```
++ critcmp main less_to_column
group                                         less_to_column                         main
-----                                         --------------                         ----
logical_aggregate_with_join                   1.01  1013.4±56.98µs        ? ?/sec    1.00  1002.7±41.73µs        ? ?/sec
logical_plan_tpcds_all                        1.00    152.8±0.76ms        ? ?/sec    1.00    153.0±1.61ms        ? ?/sec
logical_plan_tpch_all                         1.00     17.0±0.21ms        ? ?/sec    1.00     16.9±0.19ms        ? ?/sec
logical_select_all_from_1000                  1.00     18.1±0.14ms        ? ?/sec    1.04     18.9±0.08ms        ? ?/sec
logical_select_one_from_700                   1.01    820.3±9.09µs        ? ?/sec    1.00    810.2±9.31µs        ? ?/sec
logical_trivial_join_high_numbered_columns    1.01   772.3±17.02µs        ? ?/sec    1.00    762.2±6.46µs        ? ?/sec
logical_trivial_join_low_numbered_columns     1.01   762.8±20.29µs        ? ?/sec    1.00   755.9±19.28µs        ? ?/sec
physical_plan_tpcds_all                       1.00   1226.8±5.71ms        ? ?/sec    1.00   1225.2±8.60ms        ? ?/sec
physical_plan_tpch_all                        1.00     83.6±1.50ms        ? ?/sec    1.00     83.8±1.04ms        ? ?/sec
physical_plan_tpch_q1                         1.04      4.6±0.27ms        ? ?/sec    1.00      4.5±0.04ms        ? ?/sec
physical_plan_tpch_q10                        1.00      4.0±0.04ms        ? ?/sec    1.00      4.0±0.04ms        ? ?/sec
physical_plan_tpch_q11                        1.01      3.6±0.05ms        ? ?/sec    1.00      3.5±0.04ms        ? ?/sec
physical_plan_tpch_q12                        1.01      2.7±0.03ms        ? ?/sec    1.00      2.7±0.02ms        ? ?/sec
physical_plan_tpch_q13                        1.00      2.0±0.02ms        ? ?/sec    1.00      2.0±0.02ms        ? ?/sec
physical_plan_tpch_q14                        1.01      2.4±0.02ms        ? ?/sec    1.00      2.4±0.03ms        ? ?/sec
physical_plan_tpch_q16                        1.01      3.4±0.05ms        ? ?/sec    1.00      3.4±0.03ms        ? ?/sec
physical_plan_tpch_q17                        1.00      3.3±0.04ms        ? ?/sec    1.00      3.3±0.04ms        ? ?/sec
physical_plan_tpch_q18                        1.00      3.7±0.11ms        ? ?/sec    1.00      3.7±0.03ms        ? ?/sec
physical_plan_tpch_q19                        1.00      5.4±0.07ms        ? ?/sec    1.00      5.4±0.05ms        ? ?/sec
physical_plan_tpch_q2                         1.00      7.3±0.07ms        ? ?/sec    1.00      7.2±0.08ms        ? ?/sec
physical_plan_tpch_q20                        1.00      4.2±0.07ms        ? ?/sec    1.00      4.2±0.06ms        ? ?/sec
physical_plan_tpch_q21                        1.00      5.8±0.07ms        ? ?/sec    1.00      5.8±0.07ms        ? ?/sec
physical_plan_tpch_q22                        1.01      3.1±0.03ms        ? ?/sec    1.00      3.1±0.03ms        ? ?/sec
physical_plan_tpch_q3                         1.00      2.9±0.03ms        ? ?/sec    1.00      2.9±0.02ms        ? ?/sec
physical_plan_tpch_q4                         1.01      2.2±0.02ms        ? ?/sec    1.00      2.2±0.01ms        ? ?/sec
physical_plan_tpch_q5                         1.01      4.1±0.05ms        ? ?/sec    1.00      4.1±0.03ms        ? ?/sec
physical_plan_tpch_q6                         1.01  1446.6±63.18µs        ? ?/sec    1.00   1434.8±8.77µs        ? ?/sec
physical_plan_tpch_q7                         1.01      5.2±0.06ms        ? ?/sec    1.00      5.2±0.06ms        ? ?/sec
physical_plan_tpch_q8                         1.01      6.7±0.08ms        ? ?/sec    1.00      6.6±0.10ms        ? ?/sec
physical_plan_tpch_q9                         1.00      5.1±0.06ms        ? ?/sec    1.00      5.1±0.04ms        ? ?/sec
physical_select_all_from_1000                 1.00     59.0±0.52ms        ? ?/sec    1.04     61.3±0.32ms        ? ?/sec
physical_select_one_from_700                  1.03      3.6±0.03ms        ? ?/sec    1.00      3.5±0.02ms        ? ?/sec
```

</p>
</details> 

## Are there any user-facing changes?

There is a new API

### CODE REVIEW

File: datafusion/expr/src/expr.rs [hunk 1]
```diff
@@ -1333,6 +1335,46 @@ impl Expr {
...
+    /// # use datafusion_expr::col;
+    /// // For an expression `a + (b * a)`
+    /// let expr = col("a") + (col("b") * col("a"));
+    /// let refs = expr.column_refs();
+    /// // refs contains "a" and "b"
+    /// assert_eq!(refs.len(), 2);
+    /// assert!(refs.contains(&Column::new_unqualified("a")));
+    ///  assert!(refs.contains(&Column::new_unqualified("b")));
+    /// ```
+    pub fn column_refs(&self) -> HashSet<&Column> {
```
Comment (alamb) [inline datafusion/expr/src/expr.rs:1353]: The implementation of this function is quite nice now compared to the `expr_to_columns` one: https://github.com/alamb/datafusion/blob/58d0c34d77c9a5202e62b9281cdbf1046abaa096/datafusion/expr/src/utils.rs#L264-L309

File: datafusion/expr/src/utils.rs [hunk 2]
```diff
@@ -46,6 +46,7 @@ pub const COUNT_STAR_EXPANSION: ScalarValue = ScalarValue::Int64(Some(1));
 
 /// Recursively walk a list of expression trees, collecting the unique set of columns
 /// referenced in the expression
+#[deprecated(since = "40.0.0", note = "Expr::add_column_refs instead")]
```
Comment (alamb) [inline datafusion/expr/src/utils.rs:49]: this function is not used anywhere in the datafusion codebase

File: datafusion/optimizer/src/push_down_filter.rs [hunk 3]
```diff
@@ -785,7 +782,7 @@ impl OptimizerRule for PushDownFilter {
                 let mut keep_predicates = vec![];
                 let mut push_predicates = vec![];
                 for expr in predicates {
-                    let cols = expr.to_columns()?;
+                    let cols = expr.column_refs();
```
Comment (alamb) [inline datafusion/optimizer/src/push_down_filter.rs:785]: This is a pretty good example where there is no need to copy `Column`s simply to check if they are referenced.

**Comment by alamb** [comment]
I think this is now ready for review. I ran benchmarks and they show some slight improvement

File: datafusion/optimizer/src/analyzer/subquery.rs [hunk 4]
```diff
@@ -300,17 +300,16 @@ fn can_pullup_over_aggregation(expr: &Expr) -> bool {
     }) = expr
     {
         match (left.deref(), right.deref()) {
-            (Expr::Column(_), right) if right.to_columns().unwrap().is_empty() => true,
-            (left, Expr::Column(_)) if left.to_columns().unwrap().is_empty() => true,
+            (Expr::Column(_), right) if !right.any_column_refs() => true,
```
Comment (peter-toth) [inline datafusion/optimizer/src/analyzer/subquery.rs:303]: Not related to this PR, but I wonder if we could simplify this code to 
```rust
            (Expr::Column(_), right) => !right.any_column_refs(),
            (left, Expr::Column(_)) => !left.any_column_refs(),
            (Expr::Cast(Cast { expr, .. }), right)
                if matches!(expr.deref(), Expr::Column(_)) => !right.any_column_refs(),
            (left, Expr::Cast(Cast { expr, .. }))
                if matches!(expr.deref(), Expr::Column(_)) => !left.any_column_refs(),
            (_, _) => false,
```

**Review by peter-toth — APPROVED** [review]
LGTM, I have only a minor suggestion.

File: datafusion/optimizer/src/analyzer/subquery.rs (hunk 4 above)
Comment (alamb) [reply datafusion/optimizer/src/analyzer/subquery.rs]: That is an excellent idea -- done in cefb7bd0f

**Review by comphead — APPROVED** [review]
lgtm thanks @alamb

**Comment by alamb** [comment]
Thanks @comphead  and @peter-toth

**Pull request merged**